signer/migrations/0011__materialized_chains.sql

```diff
@@ -0,0 +1,148 @@
+-- This table holds the canonical chain information for Bitcoin and Stacks
+-- chains keyed by the bitcoin chain tip.
+-- 
+-- This table includes one row for each Stacks block which is anchored to a
+-- Bitcoin block in the canonical chain. The `bitcoin_chain_tip` column is used
+-- to identify the specific chain tip for which the canonical chain is
+-- materialized. If no Stacks block(s) exist for a given Bitcoin block, the
+-- `stacks_block_hash` and `stacks_block_height` columns will be `NULL` for that
+-- Bitcoin block hash.
+CREATE TABLE sbtc_signer.canonical_chains (
+    run_id INT NOT NULL,
+    bitcoin_chain_tip BYTEA NOT NULL,
+    bitcoin_block_hash BYTEA NOT NULL,
+    bitcoin_block_height BIGINT NOT NULL,
+    stacks_block_hash BYTEA,
+    stacks_block_height BIGINT
```

Comment:
No primary key? I suppose one gets created for you using the OID or something(?), but it still feels icky.

Reply:
So a PK here isn't the easiest due to the nature of the table, like this is a materialization of multiple versions of other data. A PK would essentially need to be on all columns excluding heights, which has the inherent problem that PKs can't be NULL (which the Stacks blocks can be). We have the other check in the materialization function which is sort-of like our integrity check, at least for ensuring that the same chain tip can't be accidentally materialized twice.

Otherwise, the only thing that a PK really is is a clustered unique index, i.e. defines the ordering of storage on disk plus a unique index. So what I've done is added a unique index on `canonical_chain_id, bitcoin_block_hash, stacks_block_hash`, which will allow exactly _one_ NULL stacks block per bitcoin block (i.e. no stacks blocks for that bitcoin block), and the ordering on disk will be insert-order, which will implicitly be `canonical_chain_id ASC, bitcoin_block_hash DESC, stacks_block_hash DESC` due to how the materialization function works.

One thing to consider is that the UK adds ~4MB for the table for a recent (this morning) production db dump, for a context window of 1000 (bitcoin blocks -- 32,624 stacks blocks).